src/analyze/analyze-shutdown-blame.c

```diff
@@ -0,0 +1,227 @@
+        int r;
+
+        shutdown_times = hashmap_new(&string_hash_ops);
+        if (!shutdown_times)
+                return log_oom();
+
+        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
+        if (r < 0)
+                return log_error_errno(r, "Failed to open journal: %m");
+
+        /* Look for systemd messages with units */
+        r = sd_journal_add_match(j, "SYSLOG_IDENTIFIER=systemd", 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to add journal match: %m");
+
+        /* Seek to the end and go backwards to find the most recent shutdown */
+        r = sd_journal_seek_tail(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to seek to journal tail: %m");
+
+        r = sd_journal_previous(j);
+        if (r < 0)
+                return log_error_errno(r, "Failed to iterate journal: %m");
+
+        SD_JOURNAL_FOREACH_BACKWARDS(j) {
+                const char *message = NULL, *unit = NULL;
+                usec_t timestamp;
+                _cleanup_(shutdown_time_freep) ShutdownTime *st = NULL;
+                ShutdownTime *existing;
+
+                r = sd_journal_get_realtime_usec(j, &timestamp);
+                if (r < 0)
+                        continue;
+
+                r = sd_journal_get_data(j, "MESSAGE", &data, &length);
+                if (r < 0)
+                        continue;
+
+                message = (const char*) data + STRLEN("MESSAGE=");
+                if (length <= STRLEN("MESSAGE="))
+                        continue;
+
+                r = sd_journal_get_data(j, "UNIT", &data, &length);
+                if (r < 0)
+                        continue;
+
+                unit = (const char*) data + STRLEN("UNIT=");
+                if (length <= STRLEN("UNIT="))
+                        continue;
+
+                /* Skip if this is not a stopping or stopped message */
+                if (!startswith(message, "Stopping") && !startswith(message, "Stopped"))
+                        continue;
+
+                existing = hashmap_get(shutdown_times, unit);
+
+                if (startswith(message, "Stopping")) {
+                        if (existing) {
+                                existing->start_time = timestamp;
+                                if (existing->stop_time > 0)
+                                        existing->duration = existing->stop_time - existing->start_time;
+                        } else {
+                                st = new0(ShutdownTime, 1);
+                                if (!st)
+                                        return log_oom();
+
+                                st->name = strndup(unit, strcspn(unit, "\n"));
+                                if (!st->name)
+                                        return log_oom();
+
+                                st->start_time = timestamp;
+                                st->stop_time = 0;
+                                st->duration = 0;
```

Comment:
you used new0() above, no need to initialize these fields to zero if they already are zero